idx/memory/tag_query.go

```diff
@@ -0,0 +1,404 @@
+//
+// expressions: the list of key & value pairs
+// skipEqual:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// index:       the tag index based on which to evaluate the conditions
+// not:         whether the resultSet shall be filtered by a = or != condition
+//
+func (q *TagQuery) filterByEqual(expressions []kv, skipEqual int, resultSet TagIDs, index TagIndex, not bool) {
+	for i, e := range expressions {
+		if i == skipEqual {
+			continue
+		}
+
+		for id := range resultSet {
+			if _, ok := index[e.key][e.value][id]; ok == not {
+				delete(resultSet, id)
+			}
+		}
+	}
+}
+
+// filterByMatch filters a list of metric ids by the given expressions. it is
+// assumed that the given expressions are all based on regular expressions.
+//
+// expressions: the list of key & value pairs
+// skipMatch:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// byId:        ID keyed index of metric definitions, used to lookup the tags of IDs
+// not:         whether the resultSet shall be filtered by the =~ or !=~ condition
+//
+func (q *TagQuery) filterByMatch(expressions []kv, skipMatch int, resultSet TagIDs, byId map[string]*idx.Archive, not bool) error {
+	for i, e := range expressions {
+		if i == skipMatch {
+			continue
+		}
+
+		var shortCut bool
+		var re *regexp.Regexp
+		var err error
+
+		// shortcut if pattern is ^.+, any value will match so there's no need
+		// to actually run the regular expression.
+		// (f.e. expression "key!=" will be translated to "key=~^.+")
+		if e.value == "^.+" {
+			shortCut = true
+		} else {
+			re, err = regexp.Compile(e.value)
+			if err != nil {
+				return errInvalidQuery
+			}
+		}
+
+		// cache all tags that have once matched the regular expression.
+		// this is based on the assumption that many matching tags will be repeated
+		// over multiple series, so there's no need to run the regex for each of them
+		// because once we know that a tag matches we can just compare strings
+		matchingTags := make(map[string]struct{})
+		notMatchingTags := make(map[string]struct{})
+	IDS:
+		for id := range resultSet {
+			var def *idx.Archive
+			var ok bool
+			if def, ok = byId[id.String()]; !ok {
+				// should never happen because every ID in the tag index
+				// must be present in the byId lookup table
+				CorruptIndex.Inc()
+				delete(resultSet, id)
+				continue IDS
+			}
+
+			for _, tag := range def.Tags {
+				// reduce regex matching by looking up cached matches
+				if _, ok := matchingTags[tag]; ok {
+					if not {
+						delete(resultSet, id)
+					}
+					continue IDS
+				}
+			}
+
+			for _, tag := range def.Tags {
+				// reduce regex matching by looking up cached non-matches
+				if _, ok := notMatchingTags[tag]; ok {
+					continue
+				}
+
+				tagSplits := strings.SplitN(tag, "=", 2)
```

Comment:
Maybe a micro optimization, but since you know the length of the key are searching for you could check that the character at `len(key) + 1` is an equals sign and then create the two slices directly and compare against the key slice.

Reply:
Off by one error above

Reply:
great idea: https://github.com/grafana/metrictank/commit/10ea634aba3e05786c5f4b879467c8b748c5db1d